pkg/controllers/selection/preferences.go

```diff
@@ -49,26 +49,34 @@ func NewPreferences() *Preferences {
 // scheduled to that zone. Preferences are removed iteratively until only hard
 // constraints remain. Pods relaxation is reset (forgotten) after 5 minutes.
 func (p *Preferences) Relax(ctx context.Context, pod *v1.Pod) {
-	affinity, ok := p.cache.Get(string(pod.UID))
+	spec, ok := p.cache.Get(string(pod.UID))
 	// Add to cache if we've never seen it before
 	if !ok {
-		p.cache.SetDefault(string(pod.UID), pod.Spec.Affinity)
+		// Limit cached PodSpec to only required data
+		cachedSpec := v1.PodSpec{
+			Affinity:    pod.Spec.Affinity,
+			Tolerations: pod.Spec.Tolerations,
+		}
+		p.cache.SetDefault(string(pod.UID), cachedSpec)
 		return
 	}
 	// Attempt to relax the pod and update the cache
-	pod.Spec.Affinity = affinity.(*v1.Affinity)
+	cachedSpec := spec.(v1.PodSpec)
+	pod.Spec.Affinity = cachedSpec.Affinity
+	pod.Spec.Tolerations = cachedSpec.Tolerations
 	if relaxed := p.relax(ctx, pod); relaxed {
-		p.cache.SetDefault(string(pod.UID), pod.Spec.Affinity)
+		p.cache.SetDefault(string(pod.UID), pod.Spec)
 	}
 }
 
 func (p *Preferences) relax(ctx context.Context, pod *v1.Pod) bool {
 	for _, relaxFunc := range []func(*v1.Pod) *string{
 		func(pod *v1.Pod) *string { return p.removePreferredNodeAffinityTerm(pod) },
 		func(pod *v1.Pod) *string { return p.removeRequiredNodeAffinityTerm(pod) },
+		func(pod *v1.Pod) *string { return p.toleratePreferNoScheduleTaints(pod) },
 	} {
 		if reason := relaxFunc(pod); reason != nil {
-			logging.FromContext(ctx).Debugf("Relaxing soft constraints for %s/%s since it previously failed to schedule, removing: %s", pod.Namespace, pod.Name, ptr.StringValue(reason))
+			logging.FromContext(ctx).Debugf("Relaxing soft constraints for pod since it previously failed to schedule, %s", ptr.StringValue(reason))
```

Comment:
```suggestion
			logging.FromContext(ctx).Debugf("Relaxing soft constraints since pod previously failed to schedule, %s", ptr.StringValue(reason))
```